Glide does not use SSO network stack

I have configured two accounts in the Notes app:
1. my account on Disroot.org
2. an account on my local Nextcloud server (located on my internal network)

I've noticed that for the second account (on my local server) there is no profile picture shown although it is configured in the Nextcloud account settings. It only shows the generic profile icon. I've tried with different pictures but none was displayed as profile picture.

My config is:
Server Fedora 31 (latest updates)
Nextcloud 17.0.2
Nextcloud client 2.4.0

So, i guess here we have the bad boy:

```
[02-10 19:08:07.173 18626:18642 E/InputStreamBinder]
Error during Nextcloud request
java.lang.IllegalStateException: CE_5
	at com.nextcloud.android.sso.InputStreamBinder.processRequestV2(InputStreamBinder.java:433)
	at com.nextcloud.android.sso.InputStreamBinder.performNextcloudRequestAndBodyStreamV2(InputStreamBinder.java:134)
	at com.nextcloud.android.sso.InputStreamBinder.performNextcloudRequestV2(InputStreamBinder.java:117)
	at com.nextcloud.android.sso.aidl.IInputStreamService$Stub.onTransact(IInputStreamService.java:134)
	at android.os.Binder.execTransact(Binder.java:752)
Caused by: java.lang.IllegalStateException: 302
	... 5 more
Caused by: java.lang.IllegalStateException: 
	... 5 more
```

and

```
Load failed for https://www.sever.lan/nextcloud/index.php/avatar/aval/64 with size [192x192]
class com.bumptech.glide.load.engine.GlideException: Failed to load resource
There was 1 cause:
com.nextcloud.android.sso.exceptions.NextcloudHttpRequestFailedException(HTTP request failed with HTTP status-code: 302)
 call GlideException#logRootCauses(String) for more detail
  Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetching data failed, class java.io.InputStream, REMOTE
There was 1 cause:
com.nextcloud.android.sso.exceptions.NextcloudHttpRequestFailedException(HTTP request failed with HTTP status-code: 302)
 call GlideException#logRootCauses(String) for more detail
    Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetch failed
There was 1 cause:
com.nextcloud.android.sso.exceptions.NextcloudHttpRequestFailedException(HTTP request failed with HTTP status-code: 302)
 call GlideException#logRootCauses(String) for more detail
      Cause (1 of 1): class com.nextcloud.android.sso.exceptions.NextcloudHttpRequestFailedException: HTTP request failed with HTTP status-code: 302
```

Again this `HTTP 302`... @tobiasKaminsky i had a few users (10?) which had problems with the `302`-redirect in the last week. Often this could be solved easily by reinstalling both apps and configure the accounts from scratch.

But you mentioned this should not even appear, so maybe we could add some more detailled logging in the files app in cases where a `302` is returned to find out how this constellation is even possible.

The question is: Should we put work into fixing this issue or should we dig deeper to find the actual root cause?

--- a/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnLibraryGlideModule.java
+++ b/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnLibraryGlideModule.java
@@ -9,7 +9,6 @@ import com.bumptech.glide.Registry;
 import com.bumptech.glide.annotation.GlideModule;
 import com.bumptech.glide.load.model.GlideUrl;
 import com.bumptech.glide.module.AppGlideModule;
-import com.bumptech.glide.module.LibraryGlideModule;
 
 import java.io.InputStream;
 
@@ -20,7 +19,7 @@ import java.io.InputStream;
  * annotation processor, this class will be automatically included.
  */
 @GlideModule
-public final class SingleSignOnLibraryGlideModule extends LibraryGlideModule {
+public final class SingleSignOnLibraryGlideModule extends AppGlideModule {
     @Override
     public void registerComponents(
             @NonNull Context context, @NonNull Glide glide, @NonNull Registry registry) {

--- a/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnLibraryGlideModule.java
+++ b/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnLibraryGlideModule.java
@@ -1,0 +1,29 @@
+package it.niedermann.owncloud.notes.util.glide;
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+
+import com.bumptech.glide.Glide;
+import com.bumptech.glide.Registry;
+import com.bumptech.glide.annotation.GlideModule;
+import com.bumptech.glide.load.model.GlideUrl;
+import com.bumptech.glide.module.AppGlideModule;
+import com.bumptech.glide.module.LibraryGlideModule;
+
+import java.io.InputStream;
+
+/**
+ * Registers OkHttp related classes via Glide's annotation processor.
+ *
+ * <p>For Applications that depend on this library and include an {@link AppGlideModule} and Glide's
+ * annotation processor, this class will be automatically included.
+ */
+@GlideModule
+public final class SingleSignOnLibraryGlideModule extends LibraryGlideModule {
+    @Override
+    public void registerComponents(
+            @NonNull Context context, @NonNull Glide glide, @NonNull Registry registry) {
+        registry.replace(GlideUrl.class, InputStream.class, new SingleSignOnUrlLoader.Factory(context));
+    }
+}

--- a/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnStreamFetcher.java
+++ b/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnStreamFetcher.java
@@ -1,0 +1,80 @@
+package it.niedermann.owncloud.notes.util.glide;
+
+import androidx.annotation.NonNull;
+
+import com.bumptech.glide.Priority;
+import com.bumptech.glide.load.DataSource;
+import com.bumptech.glide.load.data.DataFetcher;
+import com.bumptech.glide.load.model.GlideUrl;
+import com.nextcloud.android.sso.aidl.NextcloudRequest;
+import com.nextcloud.android.sso.api.NextcloudAPI;
+import com.nextcloud.android.sso.api.Response;
+
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Fetches an {@link InputStream} using the sso library.
+ */
+public class SingleSignOnStreamFetcher implements DataFetcher<InputStream> {
+
+    private final NextcloudAPI client;
+    private final GlideUrl url;
+
+    // Public API.
+    @SuppressWarnings("WeakerAccess")
+    public SingleSignOnStreamFetcher(NextcloudAPI client, GlideUrl url) {
+        this.client = client;
+        this.url = url;
+    }
+
+    @Override
+    public void loadData(@NonNull Priority priority, @NonNull final DataCallback<? super InputStream> callback) {
+        NextcloudRequest.Builder requestBuilder = null;
+        try {
+            requestBuilder = new NextcloudRequest.Builder()
+                    .setMethod("GET")
+                    .setUrl(url.toURL().getPath());
+            Map<String, List<String>> header = new HashMap<>();
+            for (Map.Entry<String, String> headerEntry : url.getHeaders().entrySet()) {
+                header.put(headerEntry.getKey(), Collections.singletonList(headerEntry.getValue()));
+            }
+            requestBuilder.setHeader(header);
+            NextcloudRequest nextcloudRequest = requestBuilder.build();
+            Response response = client.performNetworkRequestV2(nextcloudRequest);
+            callback.onDataReady(response.getBody());
+        } catch (MalformedURLException e) {
+            callback.onLoadFailed(e);
+        } catch (Exception e) {
+            callback.onLoadFailed(e);
+        }
+
+    }
+
+    @Override
+    public void cleanup() {
+
+    }
+
+    @Override
+    public void cancel() {
+
+    }
+
+    @NonNull
+    @Override
+    public Class<InputStream> getDataClass() {
+        return InputStream.class;
+    }
+
+    @NonNull
+    @Override
+    public DataSource getDataSource() {
+        return DataSource.REMOTE;
+    }
+}

--- a/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnUrlLoader.java
+++ b/app/src/main/java/it/niedermann/owncloud/notes/util/glide/SingleSignOnUrlLoader.java
@@ -1,0 +1,90 @@
+package it.niedermann.owncloud.notes.util.glide;
+
+import android.content.Context;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import com.bumptech.glide.load.Options;
+import com.bumptech.glide.load.model.GlideUrl;
+import com.bumptech.glide.load.model.ModelLoader;
+import com.bumptech.glide.load.model.ModelLoaderFactory;
+import com.bumptech.glide.load.model.MultiModelLoaderFactory;
+import com.google.gson.GsonBuilder;
+import com.nextcloud.android.sso.api.NextcloudAPI;
+import com.nextcloud.android.sso.exceptions.NextcloudFilesAppAccountNotFoundException;
+import com.nextcloud.android.sso.exceptions.NoCurrentAccountSelectedException;
+import com.nextcloud.android.sso.helper.SingleAccountHelper;
+
+import java.io.InputStream;
+
+/**
+ * A simple model loader for fetching media over http/https using OkHttp.
+ */
+public class SingleSignOnUrlLoader implements ModelLoader<GlideUrl, InputStream> {
+
+    private static final String TAG = SingleSignOnUrlLoader.class.getCanonicalName();
+    private final NextcloudAPI client;
+
+    // Public API.
+    @SuppressWarnings("WeakerAccess")
+    public SingleSignOnUrlLoader(@NonNull NextcloudAPI client) {
+        this.client = client;
+    }
+
+    @Override
+    public boolean handles(@NonNull GlideUrl url) {
+        return true;
+    }
+
+    @Override
+    public LoadData<InputStream> buildLoadData(
+            @NonNull GlideUrl model, int width, int height, @NonNull Options options) {
+        return new LoadData<>(model, new SingleSignOnStreamFetcher(client, model));
+    }
+
+    /**
+     * The default factory for {@link SingleSignOnUrlLoader}s.
+     */
+    // Public API.
+    @SuppressWarnings("WeakerAccess")
+    public static class Factory implements ModelLoaderFactory<GlideUrl, InputStream> {
+        private boolean isConnected = false;
+        private SingleSignOnUrlLoader loader;
+
+        /**
+         * Constructor for a new Factory that runs requests using given client.
+         */
+        public Factory(@NonNull Context context) {
+            try {
+                loader = new SingleSignOnUrlLoader(new NextcloudAPI(context, SingleAccountHelper.getCurrentSingleSignOnAccount(context), new GsonBuilder().create(), new NextcloudAPI.ApiConnectedListener() {
+                    @Override
+                    public void onConnected() {
+                        isConnected = true;
+                        Log.v(TAG, "success: init SSO-Api");
+                    }
+
+                    @Override
+                    public void onError(Exception ex) {
+                        Log.e(TAG, ex.getMessage(), ex);
+                    }
+                }));
+            } catch (NextcloudFilesAppAccountNotFoundException e) {
+                e.printStackTrace();
+            } catch (NoCurrentAccountSelectedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        @NonNull
+        @Override
+        public ModelLoader<GlideUrl, InputStream> build(MultiModelLoaderFactory multiFactory) {
+            return loader;
+        }
+
+        @Override
+        public void teardown() {
+            // Do nothing, this instance doesn't own the client.
+        }
+    }
+}

